py3status/modules/xrandr_rotate.py

```diff
@@ -62,17 +62,17 @@ def post_config_hook(self):
         self.displayed = ''
 
     def _get_all_outputs(self):
-        cmd = 'xrandr -q | grep " connected [^(]" | cut -d " " -f1'
-        return self.py3.command_output(cmd, shell=True).splitlines()
+        outputs = self.py3.command_output(['xrandr']).splitlines()
+        return [x.split()[0] for x in outputs if ' connected' in x]
```

Comment:
hey @lasers, thanks for pinging me 🙂 I tested all patches, they work well, except this change above introduces a small regression: `hide_if_disconnected` is not working anymore.

I use the following configuration:

```
xrandr {
  force_on_start = 'DP-1-2'
  output_combinations = 'DP-1-2|eDP-1'
  hide_if_single_combination = true
}

xrandr_rotate {
  screen = 'DP-1-2'
  hide_if_disconnected = true
}
```

What it does is the following: I often have my docking station plugged in the laptop, which in turn makes my external monitor often connected. Sometimes I use only laptop's screen (`eDP-1`), and sometimes I use only external screen (`DP-1-2`), but never both at the same time.

Only my external screen is rotatable, so I don't even want to see `xrandr_rotate` as a module when I'm currently using laptop's screen.

Now, this line above before was previously discovering only that screen that is currently showing output, not all connected screens, and that made a difference below in computing `selected_screen_disconnected`. The new code now returns both screens always, regardless of which one is currently showing output and which one is not outputting anything.

Output of `xrandr` when both screens are connected, but ...:
- only `DP-1-2` is used:
    ```
    Screen 0: minimum 320 x 200, current 3840 x 2160, maximum 8192 x 8192
    eDP-1 connected primary (normal left inverted right x axis y axis)
       3840x2160     60.00 +
    ...
    DP-1-2 connected 3840x2160+0+0 (normal left inverted right x axis y axis) 527mm x 296mm
       3840x2160     60.00*+
    ...
    ```
- only `eDP-1` is used:
    ```
    Screen 0: minimum 320 x 200, current 3840 x 2160, maximum 8192 x 8192
    eDP-1 connected primary 3840x2160+0+0 (normal left inverted right x axis y axis) 346mm x 194mm
       3840x2160     60.00*+
    ...
    DP-1-2 connected (normal left inverted right x axis y axis)
       3840x2160     60.00 +
    ...
    ```
- both screens are outputting something (not my usual workflow, but for completeness)
    ```
    Screen 0: minimum 320 x 200, current 3840 x 2160, maximum 8192 x 8192
    eDP-1 connected primary 3840x2160+0+0 (normal left inverted right x axis y axis) 346mm x 194mm
       3840x2160     60.00*+
    ...
    DP-1-2 connected 3840x2160+0+0 (normal left inverted right x axis y axis) 527mm x 296mm
       3840x2160     60.00*+
    ```

So this function `_get_all_outputs` should return only those outputs that are active now. As you can see, the used outputs have resolution after their name, and also in the list of all resolutions on the next line one of them has `*` symbol next to it. I guess just by looking if there is resolution after `connected` word would be the simplest way.

Reply:
@maximbaz Weird. Is it possible that it's the same behavior on the old module too? You mentioned last week or so that you don't use this anymore. Also, I ran this diff.

```
diff --git a/py3status/modules/static_string.py b/py3status/modules/static_string.py
index dbcec8c6..7d0158d0 100644
--- a/py3status/modules/static_string.py
+++ b/py3status/modules/static_string.py
@@ -16,12 +16,26 @@ class Py3status:
     """
     """
     # available configuration parameters
-    format = 'Hello, world!'
+    format = '{result}'
 
     def static_string(self):
+        cmd = 'xrandr -q | grep " connected [^(]" | cut -d " " -f1'
+        original = self.py3.command_output(cmd, shell=True).splitlines()
+
+        outputs = self.py3.command_output(['xrandr']).splitlines()
+        new = [x.split()[0] for x in outputs if ' connected' in x]
+
+        result = original == new
+
+        self.py3.log('-' * 15)
+        self.py3.log(original)
+        self.py3.log(new)
+        self.py3.log('-' * 15)
+        self.py3.log(result)
+
         return {
             'cached_until': self.py3.CACHE_FOREVER,
-            'full_text': self.py3.safe_format(self.format),
+            'full_text': self.py3.safe_format(self.format, {'result': result}),
         }
 
```

```
2017-10-11 04:10:26 INFO Module `static_string`: '---------------'
2017-10-11 04:10:26 INFO Module `static_string`: ['DP-2', 'DP-3']
2017-10-11 04:10:26 INFO Module `static_string`: ['DP-2', 'DP-3']
2017-10-11 04:10:26 INFO Module `static_string`: '---------------'
2017-10-11 04:10:26 INFO Module `static_string`: True
```
It came out same on my side. Idk. Maybe a new `xrandr_rotate` bug if you can confirm that it's the same behavior before the change... Anyhow, it'd be nice to see your diff too as it might be related to different outputs and/or xrandr versions.

Reply:
```
af7bf9a5b7824c852f0bc3435fa684e1eda0eca9 is the first bad commit
commit af7bf9a5b7824c852f0bc3435fa684e1eda0eca9
Author: Maxim Baz <git@maximbaz.com>
Date:   Wed Sep 27 14:19:15 2017 +0200

    Do not split shell commands into sequences

    The modules like `xrandr_rotate` are totally broken, because since `cmd`
    is a string, py3.command_output() will `shlex.split` it, and the command
    becomes invalid. The solution is to not split shell commands.
```
:+1:

Reply:
Ran the same diff before the first bad commit.
```
2017-10-11 04:55:33 INFO Module `static_string`: '---------------'
2017-10-11 04:55:33 INFO Module `static_string`:
['Screen 0: minimum 8 x 8, current 1920 x 1200, maximum 8192 x 8192',
 'DP-0 disconnected (normal left inverted right x axis y axis)',
 'DP-1 disconnected (normal left inverted right x axis y axis)',
 'DP-2 connected primary (normal left inverted right x axis y axis)',
 '   1920x1200     59.95 +',
 '   1920x1080     60.00  ',
 '   1680x1050     59.95  ',
 '   1600x1200     60.00  ',
 '   1280x1024     60.02  ',
 '   1280x960      60.00  ',
 '   1024x768      60.00  ',
 '   800x600       60.32  ',
 '   640x480       59.94  ',
 'DP-3 connected 1920x1200+0+0 (normal left inverted right x axis y axis) '
 '518mm x 324mm',
 '   1920x1200     59.95*+',
 '   1920x1080     60.00  ',
 '   1680x1050     59.95  ',
 '   1600x1200     60.00  ',
 '   1280x1024     60.02  ',
 '   1280x960      60.00  ',
 '   1024x768      60.00  ',
 '   800x600       60.32  ',
 '   640x480       59.94  ']
2017-10-11 04:55:33 INFO method xrandr_rotate
        returned {'full_text': '', 'instance': '', 'name': 'xrandr_rotate'}
2017-10-11 04:55:33 INFO Module `static_string`: ['DP-2', 'DP-3']
2017-10-11 04:55:33 INFO Module `static_string`: '---------------'
2017-10-11 04:55:33 INFO Module `static_string`: False
```

Reply:
I have py3status-git installed, last updated I think a few days ago. There everything works, but when I apply the patch from this PR (nothing else changes), it stops working - so I think it's a new regression. Just make the code a bit smarter and ignore those outputs that don't have resolution next to its name, and we should be good.

Is there a way to just run the module? Naive `python static_string.py` doesn't start 🙂

Regardless:

```
$ xrandr -q | grep " connected [^(]" | cut -d " " -f1
eDP-1
```

```
>>> outputs = subprocess.Popen("xrandr", stdout=subprocess.PIPE).stdout.read().splitlines()
>>> [str(x).split()[0] for x in outputs if ' connected' in str(x)]
["b'eDP-1", "b'DP-1-2"]
```

Reply:
>Is there a way to just run the module?

`git cherry-pick ff2353dc # update standalone testing` https://github.com/ultrabug/py3status/pull/1033

Reply:
Check out the old regex, it was stupid, but it said "no words between 'connected' and opening bracket", the new way just says "if there is connected" word.

Now my xrandr says:

```
eDP-1 connected primary 3840x2160+0+0 (normal left inverted right x axis y axis) 346mm x 194mm
```

(there ARE words between "connected" and opening bracket, so take it in the result)

```
DP-1-2 connected (normal left inverted right x axis y axis)
```

(there are NO words between "connected" and opening bracket, so don't take it in the result)

And the new code takes both in the result. This is a difference.

Reply:
```
3d [chris:~] % xrandr -q | grep " connected [^(]" | cut -d " " -f1
DP-2
DP-3
```
```
3d [chris:~] % python3
Python 3.6.2 (default, Jul 20 2017, 03:52:27)
[GCC 7.1.1 20170630] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import subprocess
>>> outputs = subprocess.Popen("xrandr", stdout=subprocess.PIPE).stdout.read().splitlines()
>>> [str(x).split()[0] for x in outputs if ' connected' in str(x)]
["b'DP-2", "b'DP-3"]
>>>
```

Reply:
Probably your xrandr provides a different output :) You can test on the output that I showed above. As I said, I think the easiest way would be to check if there is a resolution string between "connected" and opening bracket, if there is - take it in, otherwise don't.

Reply:
@maximbaz Good shit! Try this now... Also, I found something. This only updates once...then the rest... it's just wasting cycles... and wasting time getting `xrandr` every 10 seconds... as it's not going to do anything with it!

Reply:
Works, awesome job @lasers!

Regarding wasted cycles, you mean you would like to call `_get_all_outputs` function only once and cache its results forever? Well it won't always work, because I use `hide_if_disconnected`, when I change my output with `xrandr` module, I do want the `xrandr_rotate` to call `_get_all_outputs` and discover that I'm using a different output and either hide or show itself.

Reply:
I think here... 
```
123    if not self.displayed:
124        self.displayed = self._get_current_rotation_icon(all_outputs)
```
If it get `self.displayed`... then it'll never update the current icon... meaning you could scroll over to `V` or `H`... then leave it alone... It won't get corrected in 10 seconds... or in 10 minutes.

Reply:
Ah, I see what you mean... I guess this is by design, in current implementation you don't want it self-correct. Imagine I set `cache_timeout=1`, I left-click to scroll from `H` to `V`, but before I can right-click to confirm `V` it "get's self-corrected" back to `H`.

However this lead me to think about the following: do we really need two separate operations here, left click to scroll and right click to confirm? Maybe just make one click do the toggle? Then the shown value is always the current one, and there is nothing to "correct". Not sure if I like it or not, just a thought.

Reply:
>Imagine I set cache_timeout=1, I left-click to scroll from H to V, but before I can right-click to confirm V it "get's self-corrected" back to H.

Then it's behaving correctly, I would imagine... If you use anything external outside of `xrandr_rotate` (eg `xrandr`, `arandr`, etc...), then that icon would still not get updated. You could rotate your monitors all the way around with `1` cache_timeout and it wouldn't know... :smile:

>Imagine I set cache_timeout=1, I left-click to scroll from H to V, but before I can right-click to confirm V it "get's self-corrected" back to H.

Possible. I blame the insane `cache_timeout`. Could stay on this sensible number... And if it's really an issue, then later, we add something extra (boolean) in `on_click` to help reset / ignore updates so that does not happen again.

>left click to scroll and right click to confirm? Maybe just make one click do the toggle? 

I forget about left. Here, I use scroll up/down, then right click... Do you want something crazy? Make `V, H` happen... If I have 3 monitors, then `N N N`.... and show `N, R, L, I`.... so I can click  on left one... `L N N`... and then `L N L`... then apply. :-)

Reply:
All in all, I don't have a strong opinion for or against self-correct, if you want to try implementing it, do it in a separate review and ping me, I'll give it a test run :)

Reply:
Sure. I'll give it a go. I'd like to address multiple button issue too... https://github.com/ultrabug/py3status/issues/980 holding me back from making button PRs for both `xrandr` modules. Alright. Will ping you tmrw. Bye!